org.eclipse.xtext.ide.tests/src/org/eclipse/xtext/ide/tests/server/UriExtensionsTest.xtend

```diff
@@ -7,51 +7,263 @@
 import static org.junit.Assert.*
-import java.io.File
 
 @RunWith(XtextRunner)
 @InjectWith(TestLanguageIdeInjectorProvider)
 class UriExtensionsTest {
 
 	@Inject
-	extension UriExtensions;
+	extension UriExtensions
+
+	@Test
+	def void test_toUri01() {
+		assertEquals(
+			createURI('file://path/to/resource'),
+			'file://path/to/resource'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_02() {
+		assertEquals(
+			createURI('file:///path/to/resource'),
+			'file:///path/to/resource'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri03() {
+		assertEquals(
+			createURI('file://path with whitespaces/to/resource'),
+			'file://path with whitespaces/to/resource'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_04() {
+		assertEquals(
+			createURI('file:///path with whitespaces/to/resource'),
+			'file:///path with whitespaces/to/resource'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_05() {
+		assertEquals(
+			createURI('file:///dir/\u0424\u0443 \u0411\u0430\u0440'),
+			'file:///dir/\u0424\u0443 \u0411\u0430\u0440'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_06() {
+		assertEquals(
+			createURI('file://dir/\u0424\u0443 \u0411\u0430\u0440'),
+			'file://dir/\u0424\u0443 \u0411\u0430\u0440'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_07() {
+		assertEquals(
+			createURI('something:/path/to/resource'),
+			'something:/path/to/resource'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_08() {
+		assertEquals(
+			createURI('something://path/to/resource'),
+			'something://path/to/resource'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_09() {
+		assertEquals(
+			createURI('something:/dir/\u0424\u0443 \u0411\u0430\u0440'),
+			'something:/dir/\u0424\u0443 \u0411\u0430\u0440'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_10() {
+		assertEquals(
+			createURI('something://dir/\u0424\u0443 \u0411\u0430\u0440'),
+			'something://dir/\u0424\u0443 \u0411\u0430\u0440'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_11() {
+		assertEquals(
+			createURI('something:/path with whitespaces/to/resource'),
+			'something:/path with whitespaces/to/resource'.toUri
+		);
+	}
+
+	@Test
+	def void test_toUri_12() {
+		assertEquals(
+			createURI('something://path with whitespaces/to/resource'),
+			'something://path with whitespaces/to/resource'.toUri
+		);
+	}
 
 	@Test
+	def void test_toUriString_01() {
+		assertEquals(
+			'file:///path/to/resource',
+			URI.create('file:///path/to/resource').toUriString
+		);
+	}
+
+	@Test
+	def void test_toUriString_02() {
+		assertEquals(
+			'file://path/to/resource',
+			URI.create('file://path/to/resource').toUriString
+		);
+	}
+
+	@Test
+	def void test_toUriString_03() {
+		assertEquals(
+			'file:///dir/\u0424\u0443%20\u0411\u0430\u0440',
+			URI.create('file:/dir/\u0424\u0443%20\u0411\u0430\u0440').toUriString
+		);
+	}
+
+	@Test
+	def void test_toUriString_04() {
+		assertEquals(
+			'file://dir/\u0424\u0443%20\u0411\u0430\u0440',
+			URI.create('file://dir/\u0424\u0443%20\u0411\u0430\u0440').toUriString
```

Comment:
Same as below. I am a bit puzzled, so when the file URI has two slashes it remains the same otherwise it will be three forward slashes? Is that the logic?

Reply:
Yes, in that case the first segment is the authority. I will update the documentation.

Reply:
The other conversion is ok, according to https://en.wikipedia.org/wiki/Talk%3AFile_URI_scheme